### PR TITLE
Fb/resync

### PR DESCRIPTION
**Pull request by jekabolt**

### CODE REVIEW

File: btc/btc.go [hunk 1]
```diff
@@ -101,35 +101,35 @@ func InitHandlers(dbConf *store.Conf, coinTypes []store.CoinType, nsqAddr string
 	restoreState = db.DB(dbConf.DBRestoreState).C(dbConf.TableState)
 
 	// setup main net
-	urlMain, err := fethCoinType(coinTypes, currencies.Bitcoin, currencies.Main)
+	ctMain, err := fethCoinType(coinTypes, currencies.Bitcoin, currencies.Main)
```
Comment (nickpro) [inline btc/btc.go]: no need to use shortcuts here, let it be coinTypeMain for better readability

File: btc/btc.go [hunk 2]
```diff
@@ -145,25 +145,11 @@ func initGrpcClient(url string) (pb.NodeCommuunicationsClient, error) {
 	return client, nil
 }
 
-func fethCoinType(coinTypes []store.CoinType, currencyID, networkID int) (string, error) {
+func fethCoinType(coinTypes []store.CoinType, currencyID, networkID int) (store.CoinType, error) {
```
Comment (nickpro) [inline btc/btc.go]: fetChCoinType I guess

File: btc/handlers.go [hunk 3]
```diff
@@ -14,19 +14,35 @@ import (
...
 	"github.com/Multy-io/Multy-back/store"
-	nsq "github.com/bitly/go-nsq"
 	"gopkg.in/mgo.v2"
 	"gopkg.in/mgo.v2/bson"
 )
 
-func setGRPCHandlers(cli pb.NodeCommuunicationsClient, nsqProducer *nsq.Producer, networtkID int, wa chan pb.WatchAddress, mempool sync.Map, resync *sync.Map, btcCli *BTCConn) {
+func (btcCli *BTCConn) setGRPCHandlers(networtkID, accuracyRange int) {
+
+	var client pb.NodeCommuunicationsClient
```
Comment (nickpro) [inline btc/handlers.go]: Communications, please

File: btc/handlers.go [hunk 4]
```diff
@@ -92,31 +108,111 @@ func setGRPCHandlers(cli pb.NodeCommuunicationsClient, nsqProducer *nsq.Producer
...
+				txStore = txsDataTest
+				nsCli = btcCli.CliTest
+			}
+
+			query := bson.M{
+				"$or": []bson.M{
+					bson.M{"$and": []bson.M{
+						bson.M{"blockheight": 0},
+						bson.M{"txstatus": bson.M{"$ne": store.TxStatusTxRejectedOutgoing}},
+						bson.M{"txstatus": bson.M{"$ne": store.TxStatusTxRejectedIncoming}},
```
Comment (nickpro) [inline btc/handlers.go]: how about $nin operator?

File: btc/handlers.go [hunk 5]
```diff
@@ -92,31 +108,111 @@ func setGRPCHandlers(cli pb.NodeCommuunicationsClient, nsqProducer *nsq.Producer
...
+						},
+					}
+					_, err := txStore.UpdateAll(query, update)
+					if err != nil {
+						log.Errorf("setGRPCHandlers: cli.EventNewBlock:txStore.UpdateAll:Incoming: %s", err.Error())
+					}
+
+					// reject outcoming
+					query = bson.M{
+						"$or": []bson.M{
```
Comment (nickpro) [inline btc/handlers.go]: will $in operator do the stuff?

File: cmd/main.go [hunk 6]
```diff
@@ -60,9 +56,9 @@ func main() {
 	log.Infof("build time: %s", buildtime)
 	log.Infof("tag: %s", lasttag)
 
-	var gracefulStop = make(chan os.Signal)
+	// var gracefulStop = make(chan os.Signal)
 
-	signal.Notify(gracefulStop, syscall.SIGTERM, syscall.SIGINT, os.Interrupt)
+	// signal.Notify(gracefulStop, syscall.SIGTERM, syscall.SIGINT, os.Interrupt)
```
Comment (nickpro) [inline cmd/main.go]: let's remove commented lines

File: multy-back.go [hunk 7]
```diff
@@ -141,24 +142,32 @@ func (m *Multy) SetUserData(userStore store.UserStore, ct []store.CoinType) ([]s
...
-			// }
-
-			// if strings.Contains("err:", rp.GetMessage()) {
-			// 	log.Errorf("SetUserData:  Contains err : curID :%d netID :%d err =%s", conCred.СurrencyID, conCred.NetworkID, err.Error())
-			// 	// return servicesInfo, fmt.Errorf("SetUserData:  Contains err : curID :%d netID :%d err =%s", conCred.СurrencyID, conCred.NetworkID, err.Error())
-			// }
-			// log.Errorf("BTC ++++++++++ %v", h)
+			go func() {
+				if conCred.AccuracyRange > 0 {
+					h, err := m.userStore.FethLastSyncBlockState(conCred.NetworkID, conCred.СurrencyID)
```
Comment (nickpro) [inline multy-back.go]: fetch?

Comment (nickpro) [reply multy-back.go]: and what "h" means? please, rename to readable variable name

File: btc/handlers.go (hunk 4 above)
Comment (jekabolt) [reply btc/handlers.go]: -

Comment (jekabolt) [reply btc/handlers.go]: $nin is not suitable here

**Review by nickpro — APPROVED** [review]

File: eth/utils.go [hunk 8]
```diff
@@ -714,23 +703,26 @@ func ParseMultisigInput(tx *store.TransactionETH, networtkID int, multisigStore,
...
 
 		tx.Multisig.Owners = []store.OwnerHistory{}
-
 		log.Debugf("revokeConfirmation: %v", tx.Multisig.Input)
-		i, _ := new(big.Int).SetString(tx.Multisig.Input, 16)
 
-		sel := bson.M{"multisig.requestid": i.Int64(), "multisig.contract": contract.ContractAddress}
+		requestid, err := parseRevokeInput(tx.Multisig.Input)
+		if err != nil {
+			log.Errorf("ParseMultisigInput:revokeConfirmation: parseRevokeInput %v requestid:%v  contract:%v ", err.Error(), requestid, contract.ContractAddress)
```
Comment (nickpro) [inline eth/utils.go]: would it be better to stop function propagation (i.e return) in case of error occured?

File: eth/utils.go [hunk 9]
```diff
@@ -892,7 +884,7 @@ func parseSubmitInput(input string) (string, string) {
 		re := regexp.MustCompile(`.{64}`) // Every 64 chars
 		parts := re.FindAllString(in, -1) // Split the string into 64 chars blocks.
 
-		if len(parts) == 4 {
+		if len(parts) >= 4 {
```
Comment (nickpro) [inline eth/utils.go]: magic number 4?
please, move it to const with readable name to be aware what's going on

File: eth/utils.go [hunk 10]
```diff
@@ -902,6 +894,24 @@ func parseSubmitInput(input string) (string, string) {
 	return address, amount
 }
 
+func parseRevokeInput(input string) (int64, error) {
+	in := input[10:]
+	re := regexp.MustCompile(`.{64}`) // Every 64 chars
+	parts := re.FindAllString(in, -1) // Split the string into 64 chars blocks.
+
+	if len(parts) > 0 {
+		i, ok := new(big.Int).SetString(input, 16)
```
Comment (nickpro) [inline eth/utils.go]: let's move 16 to consts

File: store/store.go [hunk 11]
```diff
@@ -805,3 +807,53 @@ func (mStore *MongoUserStore) Close() error {
...
+		return err
+	}
+
+	count := 0
+	for _, wallet := range user.Wallets {
+		if wallet.CurrencyID == wp.CurrencyID && wallet.NetworkID == wp.NetworkID {
+			count++
+		}
+	}
+	if count < 20 {
```
Comment (nickpro) [inline store/store.go]: please, move magic number to const

File: store/store.go [hunk 12]
```diff
@@ -805,3 +807,53 @@ func (mStore *MongoUserStore) Close() error {
...
+		case currencies.Ether:
+			txs := []TransactionETH{}
+			switch wp.NetworkID {
+			case currencies.ETHMain:
+				mStore.ETHMainTxsData.Find(query).All(&txs)
+			case currencies.ETHTest:
+				mStore.ETHTestTxsData.Find(query).All(&txs)
+			}
+			if len(txs) == 0 {
+				return fmt.Errorf("maximum avalible wallets count")
```
Comment (nickpro) [inline store/store.go]: available

File: store/store.go [hunk 13]
```diff
@@ -805,3 +807,53 @@ func (mStore *MongoUserStore) Close() error {
...
+		case currencies.Bitcoin:
+			txs := []MultyTX{}
+			switch wp.NetworkID {
+			case currencies.Main:
+				mStore.BTCMainTxsData.Find(query).All(&txs)
+			case currencies.Test:
+				mStore.BTCTestTxsData.Find(query).All(&txs)
+			}
+			if len(txs) == 0 {
+				return fmt.Errorf("maximum avalible wallets count")
```
Comment (nickpro) [inline store/store.go]: available

**Review by nickpro — APPROVED** [review]

**Pull request merged**